### PR TITLE
feat: otel updates for latest spec doc

### PR DESCRIPTION
**Pull request by feywind**
Creating this for easier review on the final spec updates.

### CODE REVIEW

File: src/telemetry-tracing.ts [hunk 1]
```diff
@@ -234,84 +279,127 @@ export class PubsubSpans {
...
   }
 
   static createPublisherSpan(message: PubsubMessage, topicName: string): Span {
-    const span: Span = getTracer().startSpan(`${topicName} send`, {
+    const topicInfo = getTopicInfo(topicName);
+    const span: Span = getTracer().startSpan(`${topicName} create`, {
       kind: SpanKind.PRODUCER,
-      attributes: PubsubSpans.createAttributes({topicName}, message),
+      attributes: PubsubSpans.createAttributes(
+        getTopicInfo(topicName),
```
Comment (hongalex) [inline src/telemetry-tracing.ts]: can you reuse the value from L332 here

File: src/telemetry-tracing.ts [hunk 2]
```diff
@@ -423,15 +522,28 @@ export class PubsubSpans {
...
+    deadline?: Duration,
+    isInitial?: boolean
   ): Span | undefined {
-    const span = PubsubSpans.createChildSpan('modify ack deadline', message);
-    span?.setAttribute(
-      'messaging.gcp_pubsub.modack_deadline_seconds',
-      deadline.totalOf('second')
+    const subInfo = getSubscriptionInfo(subName);
+    const attributes = this.createAttributes(subInfo, message);
+    const span = PubsubSpans.createChildSpan(
```
Comment (hongalex) [inline src/telemetry-tracing.ts]: These RPC spans should be in their own trace as a root span rather than a child span. Maybe I'm misunderstanding what this function does?

Comment (feywind) [reply src/telemetry-tracing.ts]: Yeah, the function name is a bit confusing I guess.

```typescript
    const span = PubsubSpans.createChildSpan(
      `${subInfo.subId} ${type}`,
      message,
      undefined,
      attributes
    );
```

The `undefined` there specifies a parent span, or in this case, no parent.

Comment (feywind) [reply src/telemetry-tracing.ts]: Actually, looking at `createChildSpan`, this isn't correct. I'll fix it.

File: src/telemetry-tracing.ts [hunk 3]
```diff
@@ -234,84 +279,123 @@ export class PubsubSpans {
...
+          message.calculatedSize;
+      } else {
+        if (message.data?.length) {
+          spanAttributes['messaging.message.envelope.size'] =
+            message.data?.length;
+        }
       }
       if (message.orderingKey) {
-        spanAttributes['messaging.gcp_pubsub.message.ordering_key'] =
+        spanAttributes['messaging.gcp_pubsub.ordering_key'] =
```
Comment (hongalex) [inline src/telemetry-tracing.ts]: Actually the original was right: https://opentelemetry.io/docs/specs/semconv/messaging/gcp-pubsub/

Fixed in the doc

File: src/publisher/pubsub-message.ts [hunk 4]
```diff
@@ -55,6 +55,23 @@ export interface PubsubMessage
    * @private
    */
   publishSchedulerSpan?: tracing.Span;
+
+  /**
+   * If this is a message being received from a subscription, expose the ackId.
```
Comment (hongalex) [inline src/publisher/pubsub-message.ts]: I think expose here just means expose to the other parts of the client library and not users right?

File: src/publisher/message-queues.ts [hunk 5]
```diff
@@ -111,6 +111,7 @@ export abstract class MessageQueue extends EventEmitter {
       }
       spanMessages.forEach(m => {
         tracing.PubsubSpans.updatePublisherTopicName(m.parentSpan!, topic.name);
```
Comment (hongalex) [inline src/publisher/message-queues.ts:113]: Technically not part of this PR but why is this necessary? It looks like the same information is already passed when `createPublisherSpan` is called.

File: src/publisher/pubsub-message.ts (hunk 4 above)
Comment (feywind) [reply src/publisher/pubsub-message.ts]: Yep. I need to make a whole separate pass of converting `@private` to `@internal` since tsdoc changed that.

File: src/publisher/message-queues.ts (hunk 5 above)
Comment (feywind) [reply src/publisher/message-queues.ts:113]: Mentioned over chat, but the trouble is that, since everything is async, it's possible that the code could get to the part that sets the initial span name before the project ID is resolved. In that case, the project ID will become `{{project}}` in the final span, which isn't great. The project ID must be resolved by this point, so this just makes sure the right names are in place.

**Comment by feywind** [comment]
Merging this into the upstream PR branch.
